[FSDP][Collectives] skipping reduce_scatter when world size is 1

anshul-si opened the pull request:
**Summary:** In its current state, FSDP collectives uses cuda synchronizations and communication ops regardless of what the world size is. However, now that replicate will use FSDP, there will be instances where group size = 1 and these synchronizations and ops will be used needlessly. I have updated fsdp_collectives to skip reduce_scatter in the foreach_reduce API when world_size ‎ = 1. I have created edited a test that uses CommDebugMode to verify that the reduce_scatter has been removed. I also edited an affected test which used 1-way FSDP by verifying and changing its assert statements for CommDebugMode. I have also added a test command. 


**Test Cases**
1. pytest test/distributed/_composable/fsdp/test_fully_shard_training.py -k test_train_parity_single_worldsize1
2. pytest test/distributed/_composable/test_composability/test_2d_composability.py -k test_tp_with_fsdp_offloading


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #160563
* #160545
* #160475
* #160464
* #160457
* #160443
* #160441
* #160390
* #160344
* #160147
* __->__ #160136



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

anshul-si committed:
[FSDP][Collectives] skipping reduce_scatter when world size is 1

[ghstack-poisoned]

anshul-si committed:
Update on "[FSDP][Collectives] skipping reduce_scatter when world size is 1"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

anshul-si committed:
Update on "[FSDP][Collectives] skipping reduce_scatter when world size is 1"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

weifengpy commented on torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:
```diff
@@ -472,6 +472,11 @@ def foreach_reduce(
     """
+
+    # currently have this to avoid mypy error, should I just include the appropriate
+    # mypy ignore above?
```
what's mypy error? this means `reduce_output = torch.empty(0, device=device)` ?

anshul-si replied:
<img width="580" height="311" alt="Screenshot 2025-08-12 at 10 13 22 AM" src="https://github.com/user-attachments/assets/a3828ed0-3005-4c6d-b6a3-58ac10b98d68" />

Reduce_output is defined in two different parts of the code depending on world_size

weifengpy replied:
if it helps with possibly_undefined, maybe `assert reduce_output is not None` before using it to let mypy know? or simply adding mypy ignore?

anshul-si committed:
Update on "[FSDP][Collectives] skipping reduce_scatter when world size is 1"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

anshul-si committed:
Update on "[FSDP][Collectives] skipping reduce_scatter when world size is 1"

**Summary:** In its current state, FSDP collectives uses cuda synchronizations and communication ops regardless of what the world size is. However, now that replicate will use FSDP, there will be instances where group size = 1 and these synchronizations and ops will be used needlessly. I have updated fsdp_collectives to skip reduce_scatter in the foreach_reduce API when world_size ‎ = 1. I have created edited a test that uses CommDebugMode to verify that the reduce_scatter has been removed. I also edited an affected test which used 1-way FSDP by verifying and changing its assert statements for CommDebugMode. I have also added a test command. 


**Test Cases**
1. pytest test/distributed/_composable/fsdp/test_fully_shard_training.py -k test_train_parity_single_worldsize1
2. pytest test/distributed/_composable/test_composability/test_2d_composability.py -k test_tp_with_fsdp_offloading





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

anshul-si committed:
Update on "[FSDP][Collectives] skipping reduce_scatter when world size is 1"

**Summary:** In its current state, FSDP collectives uses cuda synchronizations and communication ops regardless of what the world size is. However, now that replicate will use FSDP, there will be instances where group size = 1 and these synchronizations and ops will be used needlessly. I have updated fsdp_collectives to skip reduce_scatter in the foreach_reduce API when world_size ‎ = 1. I have created edited a test that uses CommDebugMode to verify that the reduce_scatter has been removed. I also edited an affected test which used 1-way FSDP by verifying and changing its assert statements for CommDebugMode. I have also added a test command. 


**Test Cases**
1. pytest test/distributed/_composable/fsdp/test_fully_shard_training.py -k test_train_parity_single_worldsize1
2. pytest test/distributed/_composable/test_composability/test_2d_composability.py -k test_tp_with_fsdp_offloading





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]
